src/model/transaction/moveBlockInContentState.js

```diff
@@ -13,83 +13,272 @@
+      blocks.mergeIn(
+        originalPrevSiblingKey,
+        originalPrevSiblingBlock.merge({
+          nextSibling: originalBlockToBeMoved.getNextSiblingKey(),
+        }),
+      );
+    }
+
+    // update old next
+    if (originalNextSiblingKey) {
+      const originalNextSiblingBlock = blocks.get(originalNextSiblingKey);
+      blocks.mergeIn(
+        originalNextSiblingKey,
+        originalNextSiblingBlock.merge({
+          prevSibling: originalBlockToBeMoved.getPrevSiblingKey(),
+        }),
+      );
+    }
+
+    // update new next
+    if (newNextSiblingKey) {
+      const newNextSiblingBlock = blocks.get(newNextSiblingKey);
+      blocks.mergeIn(
+        newNextSiblingKey,
+        newNextSiblingBlock.merge({
+          prevSibling: originalBlockKey,
+        }),
+      );
+    }
+
+    // update new prev
+    if (newPrevSiblingKey) {
+      const newPrevSiblingBlock = blocks.get(newPrevSiblingKey);
+      blocks.mergeIn(
+        newPrevSiblingKey,
+        newPrevSiblingBlock.merge({
+          nextSibling: originalBlockKey,
+        }),
+      );
+    }
+
+    // update new parent
+    if (newParentKey) {
+      const newParentBlock = blocks.get(newParentKey);
+      const newParentChildrenList = newParentBlock.getChildKeys();
+      const targetBlockIndex = newParentChildrenList.indexOf(originalTargetKey);
+
+      const insertionIndex = isInsertedAfterTarget
+        ? targetBlockIndex + 1
+        : targetBlockIndex !== 0 ? targetBlockIndex - 1 : 0;
+
+      blocks.mergeIn(
+        newParentKey,
+        newParentBlock.merge({
+          children: newParentChildrenList.insert(
+            insertionIndex,
+            originalBlockKey,
+          ),
+        }),
+      );
+    }
+
+    // update block
+    blocks.mergeIn(
+      originalBlockKey,
+      originalBlockToBeMoved.merge({
+        nextSibling: newNextSiblingKey,
+        prevSibling: newPrevSiblingKey,
+        parent: newParentKey,
+      }),
+    );
+  });
+};
+
+const moveBlockInContentState = (
   contentState: ContentState,
   blockToBeMoved: BlockNodeRecord,
   targetBlock: BlockNodeRecord,
   insertionMode: DraftInsertionType,
-): ContentState {
-  invariant(
-    blockToBeMoved.getKey() !== targetBlock.getKey(),
-    'Block cannot be moved next to itself.',
-  );
-
+): ContentState => {
   invariant(insertionMode !== 'replace', 'Replacing blocks is not supported.');
 
   const targetKey = targetBlock.getKey();
-  const blockBefore = contentState.getBlockBefore(targetKey);
-  const blockAfter = contentState.getBlockAfter(targetKey);
+  const blockKey = blockToBeMoved.getKey();
+
+  invariant(blockKey !== targetKey, 'Block cannot be moved next to itself.');
 
   const blockMap = contentState.getBlockMap();
-  const blockMapWithoutBlockToBeMoved = blockMap.delete(
-    blockToBeMoved.getKey(),
-  );
-  const blocksBefore = blockMapWithoutBlockToBeMoved
+  const isExperimentalTreeBlock = blockToBeMoved instanceof ContentBlockNode;
```

Comment:
I'm curious why we use this check inestd of checking for the 'children' array like we do in other places?

Reply:
We check for `children` array only on raw checks, internally we can check as instanceof since they are already record types